### PR TITLE
[rustdoc] Remove old style files

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @QuietMisdreavus

### CODE REVIEW

**Comment by QuietMisdreavus** [comment]
Based on the discussion in https://github.com/rust-lang/rust/issues/35705 i really doubt we should be getting rid of the redirect pages. cc @rust-lang/rustdoc

**Comment by ollie27** [comment]
I don't think there's much reason to remove the macro redirects as I don't think many crates have that many macros and we don't want to break existing links.

The other redirects aren't old, they're a new style added by #35236. To remove them we would need to update anything generating links of that style like https://github.com/rust-dev-tools/rls-analysis/blob/1e0812e8595b16c086ce9a400b9a50adaf82914b/src/lib.rs#L504-L561. cc. @nrc

**Comment by GuillaumeGomez** [comment]
They're both useless and take space for nothing. :-/

**Comment by retep998** [comment]
I greatly appreciate any reduction in the number of files produced by rustdoc. Killing off redirects will substantially improve doc generation time for `winapi`, which is currently a major blocker for anyone trying to have offline documentation of their dependencies on windows.

**Comment by nrc** [comment]
> They're both useless and take space for nothing

This is not true, they allow linking to rustdoc from tools outside the compiler. Some ways to remove them:

* only generate them with a flag (which would be set by the Rust build system and docs.rs, but not when building locally)
* run rustdoc using its own web server so we can do a programmatic redirect rather than html ones
* change the existing URLs to the redirect ones
* use disambiguation pages and no namespaces by default (this would still require some redirects, but we could remove the tools ones)

**Comment by ollie27** [comment]
@nrc: what tools are using these redirects and how hard would it be for them to generate links using the current URL scheme instead?

**Comment by nrc** [comment]
> what tools are using these redirects

RLS, cargo-src, maybe some other stuff

>how hard would it be for them to generate links using the current URL scheme instead?

impossible. Basically nobody knows about the `struct` or whatever part of the URL so its impossible to synthesize. That's why the redirects exist

**Comment by GuillaumeGomez** [comment]
What about I add a flag that'll be disabled by default in order to have those files generated?

**Comment by nrc** [comment]
> What about I add a flag that'll be disabled by default in order to have those files generated?

As long as the files are there on the hosted rustdoc and docs.rs instances, that seems fine

**Comment by GuillaumeGomez** [comment]
Ok, I'll do it tomorrow.

**Comment by GuillaumeGomez** [comment]
I added the option (as stable, if that's an issue, I'll move it as unstable, but considering it allows to generate files that were already here...).

**Comment by retep998** [comment]
Would it be better to have "redirect" in the name of the flag?

**Comment by GuillaumeGomez** [comment]
I'm open to changing the name of the flag. No preference on this side.

File: src/librustdoc/lib.rs [hunk 1]
```diff
@@ -338,6 +338,11 @@ fn opts() -> Vec<RustcOptGroup> {
                        "enable-index-page",
                        "To enable generation of the index page")
         }),
+        stable("generate-old-style-files", |o| {
+            o.optflag("",
+                      "generate-old-style-files",
+                      "Option to generate files used by RLS and some other tools")
```
Comment (QuietMisdreavus) [inline src/librustdoc/lib.rs]: My suggestion:

> `--generate-redirect-pages`
>
> Generate extra pages to support legacy URLs and tool links

File: src/test/rustdoc/issue-19190.rs [hunk 2]
```diff
@@ -23,9 +23,6 @@ impl Deref for Bar {
     fn deref(&self) -> &Foo { loop {} }
 }
 
-// @has issue_19190/Bar.t.html
```
Comment (QuietMisdreavus) [inline src/test/rustdoc/issue-19190.rs]: You should either include a test that uses the flag or add it to existing tests that check for these pages.

```rust
// compile-flags:--generate-redirect-pages
```

File: src/test/rustdoc/issue-19190.rs [hunk 3]
```diff
@@ -23,9 +23,6 @@ impl Deref for Bar {
...
 // @has issue_19190/struct.Bar.html
-// @has - '//*[@id="foo.v"]' 'fn foo(&self)'
```
Comment (QuietMisdreavus) [inline src/test/rustdoc/issue-19190.rs]: This PR doesn't delete the legacy *HTML IDs*, does it? Why did you remove it from all these tests?

File: src/test/rustdoc/old-style-files.rs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![crate_name = "foo"]
```
Comment (QuietMisdreavus) [inline src/test/rustdoc/old-style-files.rs]: Remember to rename this test file if you change the name of the flag.

**Review by QuietMisdreavus — COMMENTED** [review]
Left some suggestions.

We'll need to remember to coordinate this with docs.rs when this is merged.

File: src/test/rustdoc/issue-19190.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/test/rustdoc/issue-19190.rs]: I went a bit too strongly on the tests haha.

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/test/rustdoc/issue-35169-2.rs [hunk 5]
```diff
@@ -33,19 +33,11 @@ impl DerefMut for Bar {
...
 // @has issue_35169_2/struct.Bar.html
-// @has - '//*[@id="by_ref.v"]' 'fn by_ref(&self)'
 // @has - '//*[@id="method.by_ref"]' 'fn by_ref(&self)'
-// @has - '//*[@id="by_explicit_ref.v"]' 'fn by_explicit_ref(self: &Foo)'
 // @has - '//*[@id="method.by_explicit_ref"]' 'fn by_explicit_ref(self: &Foo)'
-// @has - '//*[@id="by_mut_ref.v"]' 'fn by_mut_ref(&mut self)'
 // @has - '//*[@id="method.by_mut_ref"]' 'fn by_mut_ref(&mut self)'
-// @has - '//*[@id="by_explicit_mut_ref.v"]' 'fn by_explicit_mut_ref(self: &mut Foo)'
 // @has - '//*[@id="method.by_explicit_mut_ref"]' 'fn by_explicit_mut_ref(self: &mut Foo)'
-// @!has - '//*[@id="by_explicit_box.v"]' 'fn by_explicit_box(self: Box<Foo>)'
```
Comment (QuietMisdreavus) [inline src/test/rustdoc/issue-35169-2.rs]: This test still should be checking for the HTML IDs.

File: src/test/rustdoc/issue-35169.rs [hunk 6]
```diff
@@ -28,19 +28,11 @@ impl Deref for Bar {
...
 
-// @has issue_35169/Bar.t.html
 // @has issue_35169/struct.Bar.html
-// @has - '//*[@id="by_ref.v"]' 'fn by_ref(&self)'
 // @has - '//*[@id="method.by_ref"]' 'fn by_ref(&self)'
-// @has - '//*[@id="by_explicit_ref.v"]' 'fn by_explicit_ref(self: &Foo)'
 // @has - '//*[@id="method.by_explicit_ref"]' 'fn by_explicit_ref(self: &Foo)'
-// @!has - '//*[@id="by_mut_ref.v"]' 'fn by_mut_ref(&mut self)'
 // @!has - '//*[@id="method.by_mut_ref"]' 'fn by_mut_ref(&mut self)'
-// @!has - '//*[@id="by_explicit_mut_ref.v"]' 'fn by_explicit_mut_ref(self: &mut Foo)'
```
Comment (QuietMisdreavus) [inline src/test/rustdoc/issue-35169.rs]: Same.

File: src/test/rustdoc/structfields.rs [hunk 7]
```diff
@@ -8,8 +8,6 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-// @has structfields/Foo.t.html
-// @has - struct.Foo.html
```
Comment (QuietMisdreavus) [inline src/test/rustdoc/structfields.rs]: Since this one actually checks the *content* of the redirect pages (and these lines were added in the PR that added the redirect pages in the first place) i feel like this test should also keep them.

**Comment by QuietMisdreavus** [comment]
There's one more place where redirect pages are emitted that you aren't covering in the current PR:

https://github.com/rust-lang/rust/blob/56b0bd5b3c45f8333cc790769a7ce13a8ec3ff82/src/librustdoc/html/render.rs#L2056-L2072

**Comment by QuietMisdreavus** [comment]
I take back my last comment. After looking through the code where that's called, it turns out that those redirects are used by rustdoc itself: When an item is defined privately but re-exported publicly, its "canonical" location is still the private one. So any links that get generated to it point to the "private" location in the hierarchy, which are then redirected to the location in `cache().paths`. So those redirect pages need to stay, so as to not break rustdoc itself.

My other outstanding review comments still stand though.

**Review by QuietMisdreavus — APPROVED** [review]
With the last force-push, looks like we're good to go!

**Comment by ollie27** [comment]
I don't agree with the macro redirects being controlled by the same flag as the tool redirects. The macro redirects are there for backward compatibility and I'm not aware of any crates with so many macros that their generation is causing problems. If the tool redirects are only to be used on documentation uploaded to `doc.rust-lang.org` and `docs.rs` then they should be behind an permanently unstable option.

Also, doesn't stabilising a new command line option require documentation in the rustdoc book and a FCP?

**Comment by QuietMisdreavus** [comment]
@bors r-

I keep forgetting about the process. I'd forgotten that the flag was insta-stable and forgot about needing documentation for it. The suggestion about being perma-unstable is interesting - @nrc do you know if the tool redirects will be needed for local documentation? If they're not, then it's worth never stabilizing it.

**Comment by Dylan-DPC** [comment]
ping from triage @QuietMisdreavus @GuillaumeGomez any updates on this

**Comment by GuillaumeGomez** [comment]
I need to add a few things.

**Comment by TimNN** [comment]
Ping from triage! It looks like this PR hasn't made any progress in a while, so I'm closing it as inactive, per [our guidelines](https://forge.rust-lang.org/triage-procedure.html). Thanks for your contributions, @GuillaumeGomez, and please feel free to re-open in the future.
